src/main.c

```diff
@@ -111,60 +113,72 @@ int main(int argc, char** argv)
         return 1;
     }
 
+    physaddr_t offset = 0;
     if (args.is_linux)
     {
 	    if (!riscv32_elf_load_by_path(vm, args.bootrom, true, 0))
 	    {
 		printf("ERROR: Failed to load vmlinux ELF file.\n");
 		return 1;
 	    }
+
+	    // won't find any place for this
+	    offset = vm->mem.size - 0x8000000;
     }
-    else if (load_file_to_ram(vm, vm->mem.begin, args.bootrom) == 0)
+    else if ((offset = load_file_to_ram(vm, vm->mem.begin, args.bootrom)) == 0)
     {
         printf("ERROR: Failed to load bootrom.\n");
         return 1;
     }
 
     if (args.dtb)
     {
-	    size_t dtb_addr = vm->mem.begin + vm->mem.size - 0x8000000;
+	    physaddr_t dtb_addr = vm->mem.begin + offset;
+	    // Page align
+	    dtb_addr = (dtb_addr + gen_mask(12)) & ~gen_mask(12);
 
 	    // Explicitly set a0 to 0 as boot hart
 	    riscv32i_write_register_u(vm, REGISTER_X10, 0);
 
 	    // DTB is aligned by 2MB
-	    uint32_t dts = load_file_to_ram(vm, dtb_addr, args.dtb);
+	    physaddr_t dts = load_file_to_ram(vm, dtb_addr, args.dtb);
 	    if (dts == 0)
 	    {
 		    printf("ERROR: Failed to load DTB.\n");
 		    return 1;
 	    }
 
-	    printf("DTB loaded at: 0x%zx size: %"PRId32"\n", dtb_addr, dts);
+	    printf("DTB loaded at: 0x%"PRIxpaddr" size: %"PRIxpaddr"\n", dtb_addr, dts);
```

Comment:
Size is shown as hex without 0x prefix, not very straightforward - looks like incorrect decimal size